milli/src/search/criteria/exactness.rs

```diff
@@ -0,0 +1,353 @@
+            }
+
+            // push remainings allowed candidates as the worst valid candidates
+            candidates_array.push(allowed_candidates);
+            // reverse the array to be able to pop candidates from the best to the worst.
+            candidates_array.reverse();
+
+            Ok((all_exact_candidates, Some(Remainings(candidates_array))))
+        },
+        // pop remainings candidates until the emptiness
+        Remainings(mut candidates_array) => {
+            let candidates = candidates_array.pop().unwrap_or_default();
+            if !candidates_array.is_empty() {
+                Ok((candidates, Some(Remainings(candidates_array))))
+            } else {
+                Ok((candidates, None))
+            }
+        },
+
+    }
+}
+
+fn attribute_start_with_docids(ctx: &dyn Context, attribute_id: u32, query: &[ExactQueryPart]) -> heed::Result<Vec<RoaringBitmap>> {
+    let lowest_level = TreeLevel::min_value();
+    let mut attribute_candidates_array = Vec::new();
+    // start from attribute first position
+    let mut pos = attribute_id * 1000;
+    for part in query {
+        use ExactQueryPart::*;
+        match part {
+            Synonyms(synonyms) => {
+                let mut synonyms_candidates = RoaringBitmap::new();
+                for word in synonyms {
+                    let wc = ctx.word_level_position_docids(word, lowest_level, pos, pos)?;
+                    if let Some(word_candidates) = wc {
+                        synonyms_candidates |= word_candidates;
+                    }
+                }
+                attribute_candidates_array.push(synonyms_candidates);
+                pos += 1;
+            },
+            Phrase(phrase) => {
+                for word in phrase {
+                    let wc = ctx.word_level_position_docids(word, lowest_level, pos, pos)?;
+                    if let Some(word_candidates) = wc {
+                        attribute_candidates_array.push(word_candidates);
+                    }
+                    pos += 1;
+                }
+            }
+        }
+    }
+
+    Ok(attribute_candidates_array)
+}
+
+fn intersection_of(mut to_intersect: Vec<&RoaringBitmap>) -> RoaringBitmap {
+    match to_intersect.len() {
+        0 => RoaringBitmap::new(),
+        1 => to_intersect[0].clone(),
+        2 => to_intersect[0] & to_intersect[1],
+        _ => {
+            to_intersect.sort_unstable_by(|a, b| a.len().cmp(&b.len()).reverse());
+
+            match to_intersect.pop() {
+                None => RoaringBitmap::new(),
+                Some(candidates)  => {
+                    let mut candidates = candidates.clone();
+                    while let Some(bitmap) = to_intersect.pop() {
+                        if candidates.is_empty() { break; }
+                        candidates &= bitmap;
+                    }
+
+                    candidates
+                },
+            }
+        }
+    }
+}
```

Comment:
```suggestion
fn intersection_of(rbs: &mut [RoaringBitmap]) -> RoaringBitmap {
    rbs.sort_unstable_by_key(RoaringBitmap::len);

    let mut iter = rbs.iter();
    match iter.next() {
        Some(first) => iter.fold(first.clone(), |acc, rb| acc & rb),
        None => RoaringBitmap::new(),
    }
}
```

Reply:
You're right, this is completelly better!